src/theory/sets/cardinality_extension.h

```diff
@@ -0,0 +1,343 @@
+   * which are descendants of eqc in the cardinality graph.
+   *
+   * Argument exp contains an explanation of why the chain of children curr
+   * are descedants of . For example, say we are in context with equivalence
+   * classes:
+   *   { A, B, C^D }, { D, B ^ C,  A ^ E }
+   * We may recursively call this method via the following steps:
+   *   eqc = D, curr = {}, exp = {}
+   *   eqc = A, curr = { D }, exp = { D = B^C }
+   *   eqc = A, curr = { D, A }, exp = { D = B^C, A = C^D }
+   * after which we discover a cycle in the cardinality graph. We infer
+   * that A must be equal to D, where exp is an explanation of the cycle.
+   */
+  void checkCardCyclesRec(Node eqc,
+                          std::vector<Node>& curr,
+                          std::vector<Node>& exp);
+  /** check normal forms
+   *
+   * This method attempts to assign "normal forms" to all set equivalence
+   * classes whose types have cardinality constraints. Normal forms are
+   * defined recursively.
+   *
+   * A "normal form" of an equivalence class [r] (where [r] denotes the
+   * equivalence class whose representative is r) is a set of representatives
+   * U = { r1, ..., rn }. If there exists at least one set in [r] that has a
+   * "flat form", then all sets in the equivalence class have flat form U.
+   * If no set in E has a flat form, then U = { r } if r does not contain
+   * the empty set, and {} otherwise.
+   *
+   * A "flat form" of a set term T is the union of the normal forms of the
+   * equivalence classes that contain sets whose parent is T.
+   *
+   * In terms of the cardinality graph, the "flat form" of term t is the set
+   * of leaves of t that are descendants of it in the cardinality graph induced
+   * by the current set of assertions. Notice a flat form is only defined if t
+   * has children. If all terms in an equivalence class E with flat forms have
+   * the same flat form, then E is added as a node to the cardinality graph with
+   * edges connecting to all equivalence classes with terms that have a parent
+   * in E.
+   *
+   * In the following inference schema, the argument intro_sets is updated to
+   * contain the set of new set terms that the procedure is requesting to
+   * introduce for the purpose of forcing the flat forms of two equivalent sets
+   * to become identical. If any equivalence class cannot be assigned a normal
+   * form, then the resulting intro_sets is guaranteed to be non-empty.
+   *
+   * As an example, say we have a context with equivalence classes:
+   *   {A, D}, {C, A^B}, {E, C^D}, {C\D}, {D\C}, {A\B}, {empty, B\A},
+   * An ordered list d_oSetEqc for this context:
+   *   A, C, E, C\D, D\C, A\B, empty, ...
+   * The normal form of {empty, B\A} is {}, since it contains the empty set.
+   * The normal forms for each of the singleton equivalence classes are
+   * themselves.
+   * The flat form of each of E and C^D does not exist, hence the normal form
```

Comment:
Why E and not C^D?

Reply:
What do you mean?  Neither E nor C^D have a flat form since there are no sets that have them as parents.

Reply:
Why is the normal form of {E,C^D} is {E} and not {C^D}?

Reply:
I am assuming the first term in each set is the representative, will update the documentation to say this.